File: app/src/main/java/com/alex/morgan/bearlist/app/BearApplication.java
package com.alex.morgan.bearlist.app;

import android.app.Application;

import com.alex.morgan.bearlist.list.BearListComponent;
import com.alex.morgan.bearlist.list.DaggerBearListComponent;

public class BearApplication extends Application {

    private static BearApplication instance;

    public static BearApplication get() {
        return instance;
    }


    private AppComponent appComponent;
    private BearListComponent bearListComponent;

    @Override
    public void onCreate() {
        super.onCreate();
        instance = this;
        appComponent = DaggerAppComponent.builder().build();
    }

    public AppComponent getAppComponent() {
        return appComponent;
    }

    public BearListComponent getBearListComponent() {
        if (bearListComponent == null) {
            this.bearListComponent = DaggerBearListComponent.builder()
                    .appComponent(appComponent).build();
        }
        return bearListComponent;
    }

    public void releaseBearListComponent() {
        this.bearListComponent = null;
    }
}

File: app/src/main/java/com/alex/morgan/bearlist/app/BearApplication.kt
package com.alex.morgan.bearlist.app

import android.app.Application
import dagger.android.*
import javax.inject.Inject

open class BearApplication : Application(), HasAndroidInjector {

    @Inject
    open lateinit var androidInjector: DispatchingAndroidInjector<Any>

    override fun onCreate() {
        super.onCreate()
        instance = this
        DaggerAppComponent.builder()
            .bindContext(this)
            .build()
            .inject(this)
    }

    override fun androidInjector(): AndroidInjector<Any> = androidInjector


    companion object {

        private lateinit var instance: BearApplication
        fun get(): BearApplication {
            return instance
        }
    }
}
